Fix:Job didn't delete pods within activeDeadlineSeconds

249043822 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**What this PR does / why we need it**:


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #107226

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

249043822 commented:
/assgin @soltysh @tnozicka

249043822 commented:
/assign @soltysh @tnozicka

249043822 commented:
/ping @soltysh @tnozicka

249043822 commented:
/assign @janetkuo

249043822 commented:
Hi @janetkuo , can you help review this pr?

249043822 commented:
/ok-to-test

soltysh commented:
Thanks for your PR, here are a few questions wrt your approach/situation:

> This is due to the fact that two job workers concurrently process activeDeadlineSeconds and the other one add a higher value a new item to the queue as seen here:
> [https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L485]

How is that possible that 2 controllers are concurrently processing the same job? Theoretically an object picked from the queue should by one controller should not be synchronously picked by another while the first hasn't finished processing.

soltysh commented:
/triage accepted
/priority backlog

249043822 commented:
> Thanks for your PR, here are a few questions wrt your approach/situation:
> 
> > This is due to the fact that two job workers concurrently process activeDeadlineSeconds and the other one add a higher value a new item to the queue as seen here:
> > [https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L485]
> 
> How is that possible that 2 controllers are concurrently processing the same job? Theoretically an object picked from the queue should by one controller should not be synchronously picked by another while the first hasn't finished processing.

@soltysh 
EXP: 
1. the controller1 received a new job1, called `SyncJob()`, but it has not called `jm.updateHandler(&job)` to update job status, then job.Status.StartTime is still nil
2. then a job pod was created, enqueued the job1, the controller2 called `SyncJob()` again, this would be the case

sanchitprakash commented:
Even we are facing a similar issue with cronjobs where some of the CronJobs failed to stop even though "activeDeadlineSeconds" parameter was defined in the template. Any updates on when this issue would be resolved?

249043822 commented:
We have resolved in our private branch using this fix

soltysh commented:
> * the controller1 received a new job1, called `SyncJob()`, but it has not called `jm.updateHandler(&job)` to update job status, then job.Status.StartTime is still nil
> 
> * then a job pod was created, enqueued the job1, the controller2 called `SyncJob()` again, this would be the case

@249043822 I see what you're talking about, but I think that instead of fixing this in `updateJobStatus` we should change this condition https://github.com/kubernetes/kubernetes/blob/8ac5d4d6a92d59bba70844fbd6e5de2383a08c96/pkg/controller/job/job_controller.go#L647 to ensure we update the status even to set start time. Can you update your PR and ping me on slack pls?

249043822 commented:
> > * the controller1 received a new job1, called `SyncJob()`, but it has not called `jm.updateHandler(&job)` to update job status, then job.Status.StartTime is still nil
> > * then a job pod was created, enqueued the job1, the controller2 called `SyncJob()` again, this would be the case
> 
> @249043822 I see what you're talking about, but I think that instead of fixing this in `updateJobStatus` we should change this condition
> 
> https://github.com/kubernetes/kubernetes/blob/8ac5d4d6a92d59bba70844fbd6e5de2383a08c96/pkg/controller/job/job_controller.go#L647
> 
> to ensure we update the status even to set start time. Can you update your PR and ping me on slack pls?

@soltysh I have updated this PR with the newest master branch, but I think that changing the condition above couldn't take control of that rare case.  this is my new solution(including consider Suspend job), please review again?

soltysh commented on pkg/controller/job/job_controller.go:
```diff
@@ -627,8 +629,10 @@ func (jm *Controller) syncJob(key string) (forget bool, rErr error) {
 					// Job is continuously active.)
-					now := metav1.Now()
-					job.Status.StartTime = &now
+					if !setStartTime { // prevent StartTime to be re-set, then ActiveDeadlineSeconds lose efficacy
```
Don't change this bit, this was described in https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2232-suspend-jobs#design-details and this is working as it should.

soltysh replied:
This still applies.

soltysh reviewed (changes requested):
> @soltysh I have updated this PR with the newest master branch, but I think that changing the condition above couldn't take control of that rare case. this is my new solution(including consider Suspend job), please review again?

The current approach is breaking the assumptions for suspended jobs. I'm not in favor of this approach. Also, can you make sure to add a test case covering it?

249043822 commented:
/test pull-kubernetes-unit

249043822 commented:
/test pull-kubernetes-unit

249043822 commented:
> > @soltysh I have updated this PR with the newest master branch, but I think that changing the condition above couldn't take control of that rare case. this is my new solution(including consider Suspend job), please review again?
> 
> The current approach is breaking the assumptions for suspended jobs. I'm not in favor of this approach. Also, can you make sure to add a test case covering it?

@soltysh I have added a testcase to illustrate the suspend job design below(https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2232-suspend-jobs#design-details):

> If a Job is created with suspend: true, the StartTime field of the Job status is set only when it is resumed for the first time.

If a creation suspend job becomes resume status,  reset the StarTime may still cause ActiveDeadlineSeconds lost efficacy

soltysh commented on pkg/controller/job/job_controller.go:
```diff
@@ -651,7 +655,9 @@ func (jm *Controller) syncJob(key string) (forget bool, rErr error) {
 			job.Status.CompletedIndexes = succeededIndexes
 		}
-
+		if err := jm.ensureStartTime(&job, setStartTime); err != nil {
```
I don't think this solves your problem, what you need to do is to change the condition above which ensures that `updateHandler` is called at all.

249043822 replied:
@soltysh Can you give a sample code to illustrate what should change for the condition? The root cause for the problem is the job.Status.starttime being set twice, but I don't know how to control calling updateHandler by conditions for this problem, thanks

soltysh replied:
I'm thinking of something like:
```golang
if job.Status.Active != active || job.Status.Succeeded != succeeded || job.Status.Failed != failed || jobConditionsChanged  || setStartTime {
```
and then make sure that flag and time are properly set at start and resume.

249043822 replied:
@soltysh If using this condition to control, it maybe lose some job status update, because only starttime mistake, others job status corrrect. so I still think rechecking the starttime in updatehandler is a good catch

soltysh replied:
Right, but the problem is unless something else gets changed it won't enter that condition at all.

249043822 replied:
Yea, so if for example  job.Status.Active != active  is right, it will ignore the following conditions, the starttime may still be updated again, so I add ensureStartTime in the IF Statement{} to prevent starttime being re-set as a result of race condition

249043822 replied:
@soltysh This problem is a race condition as two job-workers (go-routines) update job status all contains similar starttime just in time. so inside a worker's sync, we could not know whether another job has just updated the job status with a short early starttime, we only can check this as nearest as writing to storage. Only get the fresh job from storage at the latest time, I could see whether job starttime need reset

249043822 replied:
@soltysh I know what you said, I will push again later, thanks

soltysh commented:
> If a creation suspend job becomes resume status, reset the StarTime may still cause ActiveDeadlineSeconds lost efficacy

Correct, but that was a conscious decision that we've heavily discussed in https://github.com/kubernetes/kubernetes/pull/98727 and then updated the KEP accordingly.

249043822 commented:
@soltysh This is actually a race condition problem  as two job-workers (goroutines) update job statuses all contains starttime just in time. so inside a worker's sync, we could not know whether another job has just updated the job status with a short early starttime, we only can check this while writing to storage. so I think doing this in updateHandler can be more accurate. But the condition below may not be neccessary, because my target is to prevent starttime set multi-times even though it won't enter that condition at all.
> if job.Status.Active != active || job.Status.Succeeded != succeeded || job.Status.Failed != failed || jobConditionsChanged  || setStartTime {

249043822 commented:
/test pull-kubernetes-integration

249043822 commented:
/test pull-kubernetes-integration

249043822 committed:
Fix:Job didn't delete pods within activeDeadlineSeconds

249043822 commented:
/cc @soltysh for the new fix, thanks

dims commented:
@249043822 will you be coming back to this? looks like @soltysh had requested a change, was that done already?

249043822 commented:
> @249043822 will you be coming back to this? looks like @soltysh had requested a change, was that done already?

Yea, I have pushed the new fix, need @soltysh review again

JaanJah commented:
What's the status of this PR? Would love to have this fix :)

249043822 commented:
This pr has stayed for a long time, do you have plan to promote this bug? @kubernetes/sig-apps-pr-reviews

soltysh reviewed (changes requested):
I'm still having a hard time understanding how in your case 2 controllers are processing the same job at the same time. If you look here: https://github.com/kubernetes/kubernetes/blob/92a1d0f84c710755a570eaf05f3e315a8c9deb1b/pkg/controller/job/job_controller.go#L191-L193
all the goroutines will share the same queue, iow. it's impossible for 2 controllers to work on a single job at the same time. 

Your current test attached to this PR works only because you're modifying the internal cache, I don't think that's the correct way to show the problem. I'd imagine running syncJob in 2 goroutines and see if it's possible to reproduce your problem with sufficient delay or with golang's race mechanism complaining about the topic. Even though that's not something that I think is going on in your cluster. 

Having thought about it a bit more I think the best possible approach rather than dealing with the fact that StartTime is being modified (in your case that difference was only of 1 second) is to ensure that the job will get requeued again even when the other controller changes the start time.

249043822 commented:
@soltysh Thanks for your review, the case is as follows that i explained before: 
1. the controller1 received a new job1, called SyncJob(), but it has not called jm.updateHandler(&job) to update job status, then job.Status.StartTime is still nil
2. then a job pod was created, enqueued the job1, the controller2 called SyncJob() again, which would set the job.Status.StartTime again, but the delay queue trigger time would not be reset to a later time(see the issue for details)
3.  when the delay queue trigger time comes, the current time which mimus job starttime is still less than the activedeadline, that would be the case

249043822 commented:
I our environment(which is not have high performance), we can reproduce the bug easily using the job template in the issue, maybe i can add a e2e test case for test coverage  to prevent regression

249043822 commented:
/retest pull-kubernetes-e2e-gce-ubuntu-containerd

249043822 commented:
test pull-kubernetes-e2e-gce-ubuntu-containerd

249043822 commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

249043822 commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

249043822 commented:
/retest

249043822 commented:
/retest

249043822 commented:
/retest

249043822 commented:
/retest

alculquicondor commented:
There is a much simpler solution: call `jm.queue.AddAfter` every time there is a syncJob

FillZpp commented on pkg/controller/job/job_controller.go:
```diff
@@ -742,13 +749,19 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (forget bool, rEr
-			klog.V(4).Infof("Job %s has ActiveDeadlineSeconds will sync after %d seconds",
-				key, *job.Spec.ActiveDeadlineSeconds)
-			jm.queue.AddAfter(key, time.Duration(*job.Spec.ActiveDeadlineSeconds)*time.Second)
+		hasSetStartTime, startTime := jm.hasJobStartTimeSet(key)
```
What if someone deletes an existing job and then creates a new job with the same namespace/name immediately?

The `jm.removeJobStartTime(key)` above may not be called because of the quick recreate, and here it will get the wrong startTime of the previous job.

249043822 commented:
See: #110294
/close
